library/bdBank/Installer.php

```diff
@@ -67,14 +78,25 @@ class bdBank_Installer
             'field' => 'bdbank_money',
             'checkQuery' => 'SHOW COLUMNS FROM `xf_user` LIKE \'bdbank_money\'',
             'addQuery' => 'ALTER TABLE `xf_user` ADD COLUMN `bdbank_money` DECIMAL(13,4) DEFAULT \'0\'',
+            'modifyQuery' => 'ALTER TABLE `xf_user` MODIFY COLUMN `bdbank_money` DECIMAL(13,4) DEFAULT \'0\'',
             'dropQuery' => 'ALTER TABLE `xf_user` DROP COLUMN `bdbank_money`',
         ),
+        array(
+            'table' => 'xf_user',
+            'tableCheckQuery' => 'SHOW TABLES LIKE \'xf_user\'',
+            'field' => 'bdbank_credit',
```

Comment:
Field should be something adjustment I think? Credit seems wrong...

Reply:
credit is the negated sum of adjustments anh.  
Also this makes more sense for addons like TinhteRewardStore to add credit for each transaction made. For example, exchanging 5 bdbank_money for a reward will grant 5 credit points

Reply:
> credit = outgoingAdjustment - incomingAdjustment anh, so if adjustment = -5, credit will be 5

Sounds too complicated eh. How about keeping track of adjustment the normal way (income - outgoing, adjustment will be -5) and use `money - adjustment` (money minus adjustment) instead of `money + credit` (money plus credit) for user group promotion?

Reply:
Because I think `credit` is more generic anh.  
It is more straight-forward and makes more sense for other addons like TinhteRewardStore to extend the calculation of `credit`, compared to `adjustment`.

Reply:
Other add-ons still need to insert record into adjustment table eh? So they have to understand the adjust concept anyway...

Oh, you plan to extend `_calculateAdditionalCredit` in those add-on instead of adding adjustment directly? That means there will be more tables involved in calculating each user point? It's currently quite complicated already, I'm not sure it's a good idea to stretch it even more.

Reply:
yes anh, I plan to extend `_calculateAdditionalCredit`  in other addons. This only affects users' promotion so I think it's fine?  
I think it's kinda wrong for TinhteRewardStore to create an adjustment record for each of its transaction

Reply:
What is the implementation plan for the other add-on? 

## A. User purchase flow

1. User decides to buy something
1. Add-on insert transaction record using `bdBank_Model_Personal::transfer` (e.g. transaction from user -> bank, amount=100p) -> bdBank update `bdbank_money` field
1. Add-on insert purchase record in its own table and update `bdbank_credit` by itself  (e.g. bdbank_credit = bdbank_credit + 100)

## B. User rebuild flow

1. Add-on override `_calculateAdditionalCredit` to return total purchase amount from its own table

A3 sounds pretty crazy to me.

Reply:
Yes the above is what I'm planning to do.  
Could you clarify your concerns anh? :<  

In A3, the logic to update `bdbabnk_credit` might be implemented by bdBank itself.
Does it make sense to:
* make a new transaction type, say `bdBank_Model_Bank::TYPE_CREDITABLE`
* when a transaction of this new type is made, the transaction's amount is negated and added to from_user's credit

Reply:
If we go with the `TYPE_CREDITABLE`, the other addons would also **not** need to override `_calculateAdditionalCredit`

Reply:
If you let bdBank update `bdbank_credit` then the other add-on doesn't need to care about credit / adjustment -> we shouldn't introduce 2 completely new concepts in this PR (adjustment and credit). My point is: stick to one, either use adjustment for everything or just use credit everywhere. Currently it's quite confusing, adjustment in some and credit in others...

Regarding the new transaction type, I think that's great. Let's do that.

> when a transaction of this new type is made, the transaction's amount is negated and added to from_user's credit

I think we should do 3 things:

- Insert transaction record
- Insert adjustment record
- Update xf_user fields

All in the same db transaction.

Reply:
hmm I don't want to stretch the meaning of adjustments. It is weird to insert a transaction and insert an adjustment for that new transaction at the same time.  

I think the meaning of adjustment and credit can be summarized like this:
* An adjustment is a modification to a transaction's amount and is made by the system. It affects users' bdbank_money
* Credits are the points that are only considered when evaluating a user's rank. Those points are granted by adjustments or creditable transactions. Credits do not affect bdbank_money

Reply:
Using credit for everything might make more sense than using adjustment for everything, in terms of meaning.  
Let me try this approach.

Reply:
Ah no,
* using credit for everything: cannot distinguish between credits made adjustments and credits made by creditable transactions. The former is counted towards bdbank_money why the latter is not.
* using adjustment for everything: cannot distinguish between adjustments made by actual system adjustments and adjustments made by creditable transactions. The former is counted towards bdbank_money why the latter is not.

-> Need to distinguish credit and adjustment for bdbank_money calculation

Reply:
Yeah, if merge two concepts together, a slight change is required during rebuilding:

- Currently (in this PR): rebuilder will insert new adjustment record
- Required change: rebuilder should add insert 1 transaction record and 1 adjustment record

Example: post bonus from 10 -> 5.

- Currently (in this PR): rebuilder will insert adjustment record of -5 (target 5 - current 10 = -5)
- Required change: rebuilder will insert transaction -5 **and** adjustment +5.

Benefits:

- Only one new concept
- Simplify `bdbank_money` calculation: just sum(amount) of all transaction
- Easy to calculate `bdbank_credit` (or `bdbank_adjustment`, up to you): sum(amount) of all adjustment

Thinking about this, is it possible to use transaction table for adjustment too? The adjustment amount is always negative of the transaction amount, if we set `transaction_type=adjustment` and take it into account while calculating `bdbank_money` and `bdbank_credit`, that will be fast.

Reply:
implemented the above approach via https://github.com/xfrocks/bdBank/pull/4/commits/360985f6504bb8c1aba0551adbc73e119cffaaa5: 
* track adjustments using `transaction` table
* create `credit` record for every adjustment/creditable transaction, by using a DB trigger. Without the trigger, we cannot know which credit is made for which transaction. One drawback is that it will be difficult  to update the trigger definition, we will probably need to lock bdbank_transaction table to do so.
* update `xf_user.bdbank_credit` every time a `credit` record is inserted, by using a DB trigger. Trigger is necessary because of the above trigger. Besides, the trigger reduces code complexity and ensures we do not miss updating bdbank_credit
* rebuild `xf_user.bdbank_money` like before
* rebuild `xf_user.bdbank_credit` = SUM(amount) of credit records of that user

Reply:
Remind me why you still need the `xf_bdbank_credit` table?

Reply:
The triggers could be removed if `lastInsertId` is thread-safe, but I'm not sure if it is?

Reply:
It is because everything is in a transaction.

However, I don't think you need that table at all... Why all the hassle?

Reply:
ah right...

Reply:
updated to use transactions to track everything: https://github.com/xfrocks/bdBank/compare/360985f..dd1fa99

Reply:
Hmm just found 1 issue:
1. User makes 1 post and is awarded 1 point (transaction 1 with amount 1) -> bdbank_money = 1
2. Admin adjusts the bonus for posts to 3 points
3. Admin rebuilds bonuses, which creates new ADJUSTMENT transaction 2 with amount 2. Note that at this point bdbank_money is still 1
4. User deletes the post -> system tries to revert transactions 1 and 2, but will only be able to revert one of them, because user doesn't have sufficient money

Possible solutions:
1. Stop trying to make a dummy transaction to validate the reverse (https://github.com/xfrocks/bdBank/pull/4/files#diff-f555f4f5a103abc1dca15df0eb138b7dR381)
2. Make sure bdbank_money is updated within the same transaction with bonus rebuilding

Reply:
After step (3), user values will be money=1, credit=2 -- is this correct?

If it is, when step (4) happens, reverting t1 will change money=0, credit=2 then reverting t2 will change money=0, credit=0 -- I think?

> Make sure bdbank_money is updated within the same transaction with bonus rebuilding

This cannot be done because rebuilding is already very slow. If we update `xf_user` during this time, it will be even slower...

Reply:
> After step (3), user values will be money=1, credit=2 -- is this correct?
Yes anh
> If it is, when step (4) happens, reverting t1 will change money=0, credit=2 then reverting t2 will change money=0, credit=0 -- I think?
Reverting t2 would be ignored because making the dummy transaction would fail. The solution 1 above tries to solve this

Reply:
> Reverting t2 would be ignored because making the dummy transaction would fail.

I think it's a bug in the test-transaction code because it didn't take into account transaction type. e.g. only verify if type=system or type=user?

Reply:
Not sure if it's safe to verify for certain types only? What is the reason for making a test transaction when reverting anh, and it is a good idea to ignore the fail test transactions?

Reply:
I have taken a closer look, it is ignoring transaction exception already eh?

Reply:
Yeah it is ignoring anh, but I'm questioning whether ignoring is good

Reply:
I don't recall the reasoning but if it was ignoring already then let keep it this way, at least for now.

Reply:
Atm the exception is ignored but the the transaction won't be reversed.  
So should I make the transaction testing ignore TYPE_ADJUSTMENT transactions?

Reply:
Ah... I see.

No, don't ignore by type for now. Adjustment will affect `bdbank_money` and should be tested by the same logic.

> Atm the exception is ignored but the the transaction won't be reversed.

Sounds dangerous to change this logic, it may have some serious consequence that we don't know for now. Let's create a new issue to look into it later and leave it out of this PR.

Reply:
Ok anh, I will create the issue.  
Please help review this PR again nha